packages/muelu/src/Transfers/GeneralGeometric/MueLu_RegionRFactory_def.hpp

```diff
@@ -0,0 +1,1015 @@
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact
+//                    Jonathan Hu       (jhu@sandia.gov)
+//                    Andrey Prokopenko (aprokop@sandia.gov)
+//                    Ray Tuminaro      (rstumin@sandia.gov)
+//
+// ***********************************************************************
+//
+// @HEADER
+#ifndef MUELU_REGIONRFACTORY_DEF_HPP
+#define MUELU_REGIONRFACTORY_DEF_HPP
+
+#include <Xpetra_Matrix.hpp>
+#include <Xpetra_CrsGraphFactory.hpp>
+
+// #include "MueLu_PFactory.hpp"
+// #include "MueLu_FactoryManagerBase.hpp"
+#include "MueLu_Monitor.hpp"
+#include "MueLu_MasterList.hpp"
+
+#include "MueLu_RegionRFactory_decl.hpp"
+
+namespace MueLu {
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  RCP<const ParameterList> RegionRFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  GetValidParameterList() const {
+    RCP<ParameterList> validParamList = rcp(new ParameterList());
+
+// #define SET_VALID_ENTRY(name) validParamList->setEntry(name, MasterList::getEntry(name))
+//     SET_VALID_ENTRY("region: build coarse coordinates");
+// #undef  SET_VALID_ENTRY
+
+    validParamList->set<RCP<const FactoryBase> >("A",                            Teuchos::null,
+                                                 "Generating factory of the matrix A");
+    validParamList->set<RCP<const FactoryBase> >("numDimensions",                Teuchos::null,
+                                                 "Number of spacial dimensions in the problem.");
+    validParamList->set<RCP<const FactoryBase> >("lNodesPerDim",                 Teuchos::null,
+                                                 "Number of nodes per spatial dimension on the fine grid.");
+    validParamList->set<RCP<const FactoryBase> >("Nullspace",                    Teuchos::null,
+                                                 "Fine level nullspace used to construct the coarse level nullspace.");
+    validParamList->set<RCP<const FactoryBase> >("Coordinates",                  Teuchos::null,
+                                                 "Fine level coordinates used to construct piece-wise linear prolongator and coarse level coordinates.");
+    validParamList->set<bool>                   ("keep coarse coords",           false, "Flag to keep coordinates for special coarse grid solve");
+
+    return validParamList;
+  } // GetValidParameterList()
+
+  template <class Scalar, class LocalOrdinal, class GlobalOrdinal, class Node>
+  void RegionRFactory<Scalar, LocalOrdinal, GlobalOrdinal, Node>::
+  DeclareInput(Level& fineLevel, Level& /* coarseLevel */) const {
+    const ParameterList& pL = GetParameterList();
+
+    Input(fineLevel, "A");
+    Input(fineLevel, "numDimensions");
+    Input(fineLevel, "lNodesPerDim");
+    Input(fineLevel, "Nullspace");
+
+    if(pL.get<bool>("interp: build coarse coordinates")) {
```

Comment:
Where is this parameter defined as valid? `GetValidParameterList()` has some commented code along the lines of `SET_VALID_ENTRY("region: build coarse coordinates");`, but it's commented and uses a slightly different name for the parameter list entry.

Reply:
Yeah at this point my test does not implement the request/release mechanism properly and having that parameter enabled would lead to a runtime error being thrown by MueLu. While this is understandable it makes it complicated to develop and debug the algorithm. I will look into refining the test and turning back on the coarse coordinates calculation.

Reply:
So after further thinking, since this factory focuses on linear interpolation it always requires coordinates to be present so I removed completely that boolean parameter.